### PR TITLE
trying to warn about too big segment depth change

### PR DESCRIPTION
**Pull request by GitPaean**
It will give warning when there is segments that are defined with depth change larger than the segment length. 

Something like the following,
```
Warning:  Segment 3 of well PROD01 has a depth change of 26 meters, while it has a length of 25 meters, which is unphysical.

Warning:  Segment 4 of well PROD01 has a depth change of 33 meters, while it has a length of 25 meters, which is unphysical.
```

It is using the internal unit system `SI`, it might be better to use the original unit from the DATA file, while it might be more involved. Please suggest.

### CODE REVIEW

**Comment by GitPaean** [comment]
probably we should warn based on keyword level, obstacle is that what if the information is not completed for now.  Will look into it. probably can be done after each keyword input, to see if found anything not right onsite. 

The goal is to give warning once per parsing of the schedule, while not repeated.

**Comment by GitPaean** [comment]
I am marking the PR to be ready for review now, and @vkip , please suggest whether it looks okay to you. It is an issue from some while ago, sorry for the delay.

**Comment by GitPaean** [comment]
jekins build this please

File: opm/input/eclipse/Schedule/MSW/WellSegments.cpp [hunk 1]
```diff
@@ -616,6 +618,22 @@ namespace Opm {
...
+        for (const auto& segment : this->m_segments) {
+            const int segment_number = segment.segmentNumber();
+            if (segment_number == 1) {
+                continue; // not check the top segment for now
+            }
+            const double segment_length = this->segmentLength(segment_number);
+            const double segment_depth_change = this->segmentDepthChange(segment_number);
+            if (std::abs(segment_depth_change) > 1.001 * segment_length) { // 0.1% tolerance for comparison
+                const std::string msg = fmt::format(" Segment {} of well {} has a depth change of {} meters,"
+                                                    " while it has a length of {} meters, which is unphysical.",
```
Comment (bska) [inline opm/input/eclipse/Schedule/MSW/WellSegments.cpp:631]: Talking about "meters" is going to look somewhat out of place if the run uses FIELD units (lengths in feet).  If there's no (easy) way to get a `UnitSystem` object here, then I suggest we rephrase it to use relative measures instead, e.g.,

> Depth change in segment {} of well {} exceeds total segment length by a factor of {}.

and then pass `abs(segment_depth_change) / segment_length` as the "factor".

**Review by bska — COMMENTED** [review]
It makes sense to me to warn about this&ndash;we might even consider aborting the simulation run&ndash;but the diagnostic message itself could be improved.  At minimum, the word "while" should be replaced by "but".

File: opm/input/eclipse/Schedule/MSW/WellSegments.cpp (hunk 1 above)
Comment (GitPaean) [reply opm/input/eclipse/Schedule/MSW/WellSegments.cpp:631]: Let me investigate a little bit how to get the Unit System here, otherwise, I will go with your suggestion regarding using a factor instead.

File: opm/input/eclipse/Schedule/MSW/WellSegments.cpp [hunk 2]
```diff
@@ -43,6 +44,7 @@
 #include <unordered_set>
 #include <utility>
 #include <vector>
+#include <iostream>
```
Comment (bska) [inline opm/input/eclipse/Schedule/MSW/WellSegments.cpp:47]: This header (`<iostream>`) appears to be unused.  In that case, I'd prefer that we not include it.

File: opm/input/eclipse/Schedule/MSW/WellSegments.cpp [hunk 3]
```diff
@@ -62,8 +64,8 @@ namespace Opm {
     }
 
 
-    WellSegments::WellSegments(const DeckKeyword& keyword) {
-        this->loadWELSEGS(keyword);
+    WellSegments::WellSegments(const DeckKeyword& keyword, const UnitSystem& unit_system) {
+        this->loadWELSEGS(keyword, unit_system);
```
Comment (bska) [inline opm/input/eclipse/Schedule/MSW/WellSegments.cpp:68]: I don't see a better way right now, but having the `UnitSystem` be a parameter to the constructor makes it seem like the `WellSegments` object might plausibly retain a reference (or a full copy) of the `unit_system` argument.  This is something to keep in mind for future work, and I'm not saying we should do anything about it right now.

**Review by bska — COMMENTED** [review]
This is starting to look good to me.  Just one small issue concerning the `<iostream>` header remaining.

File: opm/input/eclipse/Schedule/MSW/WellSegments.cpp (hunk 2 above)
Comment (GitPaean) [reply opm/input/eclipse/Schedule/MSW/WellSegments.cpp]: It is leftover when debugging, removed it now.

**Comment by GitPaean** [comment]
I added unit name mapping so that we can output meters, feet and cm, but I think there might be more coding opinions join about how to make it nice, so please have a look.

File: opm/input/eclipse/Schedule/MSW/WellSegments.cpp (hunk 3 above)
Comment (GitPaean) [reply opm/input/eclipse/Schedule/MSW/WellSegments.cpp:67]: > object might plausibly retain a reference (or a full copy) 

I thought about that, but the only usage of the UnitSystem is for the purpose of outputting the message from this PR, so I decided to pass it around.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/input/eclipse/Schedule/MSW/WellSegments.cpp (hunk 3 above)
Comment (GitPaean) [reply opm/input/eclipse/Schedule/MSW/WellSegments.cpp:67]: I can do that. Let me update it.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me now and I appreciate your efforts.  I'll merge into master.

**Comment by GitPaean** [comment]
Just leave a note here, when debugging another issue, I found that we can also do the consistency checking in the function 
```
readOnIORank
```
where the information should be complete for sure. 
other consistency checking can also be done here.

**Comment by gdfldm** [comment]
@GitPaean does this check allow for horizontal or deviated segments?

**Comment by GitPaean** [comment]
> @GitPaean does this check allow for horizontal or deviated segments?

not totally following your comments but it should be generic and not introducing any problems for different types of segments.
